include/gsl/gsl-lite.hpp

```diff
@@ -1540,32 +1560,54 @@ class not_null
     }
 #endif // gsl_HAVE( RVALUE_REFERENCE )
 
-#if gsl_HAVE( FUNCTION_REF_QUALIFIER )
-# if gsl_CONFIG( NOT_NULL_TRANSPARENT_GET )
-    gsl_api gsl_constexpr14 element_type &  get() const &  { return checked_ptr().get(); }
-    gsl_api gsl_constexpr14 element_type && get() &&       { return std::move(checked_ptr()).get(); }
-# else
-    gsl_api gsl_constexpr14 get_result_t get() const { return checked_ptr(); }
-# endif
-
-    gsl_api gsl_constexpr14 operator T() const & { return checked_ptr(); }
-    gsl_api gsl_constexpr14 operator T() &&      { return std::move(checked_ptr()); }
+#if gsl_CONFIG( NOT_NULL_TRANSPARENT_GET )
+    gsl_api gsl_constexpr14 element_type* get() const { return checked_ptr().get(); }
+#else
+    gsl_api gsl_constexpr14 get_result_t  get() const { return checked_ptr(); }
+#endif
 
-    gsl_api gsl_constexpr14 T const & operator->() const & { return checked_ptr(); }
-    gsl_api gsl_constexpr14 T &&      operator->() &&      { return std::move(checked_ptr()); }
+    // Ideally, we would convert to either `T const &` or `T &&` so we would permit a move but still avoid potentially unnecessary copies.
+    // Unfortunately, that would render code like `std::unique_ptr<T> p = not_null<std::unique_ptr<T>>(...);` ambiguous.
+    // We choose to always convert to `T` instead if function ref qualifiers are available, expecting that the cost of the copy will usually be
+    // alleviated by move semantics.
+    // This implies we cannot convert to const refs of move-only types; but that scenario should be rare.
```

Comment:
I have a hard time understanding this comment. Especially this part:

> We choose to always convert to `T` instead if function ref qualifiers are available

However, we're NOT converting to T (as in the parameter of not_null), but U (a type T is convertible to). If I interpret this correctly, you used "T" here as a wildcard type, but it's confusing because T is a name already used for something else in this context.

Reply:
I'll try to rephrase.

```c++
// We want an implicit conversion operator that can be used to convert from both lvalues (by
// const reference or by copy) and rvalues (by move). So it seems like we could define
//     template< class U, gsl_REQUIRE_A(...) >
//     operator U const &() const & { return checked_ptr(); }
//     template< class U, gsl_REQUIRE_A(...) >
//     operator U&&() && { return std::move( checked_ptr() ); }
// However, having two conversion operators with different return types renders the assignment
// operator of the result type ambiguous:
//     not_null<unique_ptr<T>> p( ... );
//     unique_ptr<U> q;
//     q = std::move( p ); // ambiguous
// To avoid this ambiguity, we have both overloads of the conversion operator return `U`
// rather than `U const &` or `U &&`. This implies that converting an lvalue always induces
// a copy, which can cause unnecessary copies or even fail to compile in some situations:
//     not_null<shared_ptr<T>> sp( ... );
//     shared_ptr<U> const & rs = sp; // unnecessary copy
//     unique_ptr<U> const & ru = p; // error: cannot copy `unique_ptr<T>`
// However, these situations are rather unusual, and the following, more frequent situations
// remain unimpaired:
//     shared_ptr<U> vs = sp; // no extra copy
//     unique_ptr<U> vu = std::move( p );
```

Is that better? Also much longer, though.

Reply:
Much, much better. :)